datafusion/core/src/datasource/file_format/mod.rs

```diff
@@ -87,6 +98,277 @@ pub trait FileFormat: Send + Sync + fmt::Debug {
+    }
+
+    fn poll_flush(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<std::result::Result<(), Error>> {
+        Pin::new(&mut self.get_mut().writer).poll_flush(cx)
+    }
+
+    fn poll_shutdown(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<std::result::Result<(), Error>> {
+        Pin::new(&mut self.get_mut().writer).poll_shutdown(cx)
+    }
+}
+
+impl<W: AsyncWrite + Unpin + Send> FileWriterExt for AsyncPut<W> {}
+
+/// An extension trait for `AsyncWrite` types that adds an `abort_writer` method.
+pub trait FileWriterExt: AsyncWrite + Unpin + Send {
+    /// Aborts the writer and returns a boxed future with the result.
+    /// The default implementation returns an immediately resolved future with `Ok(())`.
+    fn abort_writer(&self) -> Result<BoxFuture<'static, Result<()>>> {
+        Ok(async { Ok(()) }.boxed())
+    }
+}
+
+/// A wrapper around an `AsyncWrite` type that provides multipart upload functionality.
+pub struct AsyncPutMultipart<W: AsyncWrite + Unpin + Send> {
+    writer: W,
+    /// A shared reference to the object store
+    store: Arc<dyn ObjectStore>,
+    multipart_id: MultipartId,
+    location: Path,
+}
+
+impl<W: AsyncWrite + Unpin + Send> AsyncPutMultipart<W> {
+    /// Create a new `AsyncPutMultipart` instance with the given writer, object store, multipart ID, and location.
+    pub fn new(
+        writer: W,
+        store: Arc<dyn ObjectStore>,
+        multipart_id: MultipartId,
+        location: Path,
+    ) -> Self {
+        Self {
+            writer,
+            store,
+            multipart_id,
+            location,
+        }
+    }
+}
+
+impl<W: AsyncWrite + Unpin + Send> AsyncWrite for AsyncPutMultipart<W> {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<std::result::Result<usize, Error>> {
+        Pin::new(&mut self.get_mut().writer).poll_write(cx, buf)
+    }
+
+    fn poll_flush(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<std::result::Result<(), Error>> {
+        Pin::new(&mut self.get_mut().writer).poll_flush(cx)
+    }
+
+    fn poll_shutdown(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<std::result::Result<(), Error>> {
+        Pin::new(&mut self.get_mut().writer).poll_shutdown(cx)
+    }
+}
+
+impl<W: AsyncWrite + Unpin + Send> FileWriterExt for AsyncPutMultipart<W> {
+    fn abort_writer(&self) -> Result<BoxFuture<'static, Result<()>>> {
+        let location = self.location.clone();
+        let multipart_id = self.multipart_id.clone();
+        let store = self.store.clone();
+        Ok(Box::pin(async move {
+            store
+                .abort_multipart(&location, &multipart_id)
+                .await
+                .map_err(DataFusionError::ObjectStore)
+        }))
+    }
+}
+
+/// A wrapper around an `AsyncWrite` type that provides append functionality.
```

Comment:
You are right. These types were exactly same (we missed it and tried to replicate file writer mode enum). However, after suggestion of @tustvold I have changed the abort implementation of the append case.  Hence this is no longer the case. However, these type have common code, I will try to increase co reuse between these types so that the places they differ is more clear.